3D Scene system upgrade.

yldrefruz opened the pull request:
# Description

As you know 3D Scene's mesh isn't using a high resolution texture. This PR solves this by adding different options of texture qualities from 512x512 to 8192x8192. Additionally adds: shadow map changes, sun color,wireframe view and an insignificant typo fix.

A 512x512 skin:
![image](https://github.com/Azgaar/Fantasy-Map-Generator/assets/30903352/196f20cf-a053-43de-9a27-5c62619731dd)


A 4096x4096 skin:
![image](https://github.com/Azgaar/Fantasy-Map-Generator/assets/30903352/991328e3-732d-4348-a636-6727eb0605d5)



# Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Refactoring / style
- [ ] Documentation update / chore
- [ ] Other (please describe)

# Versioning

- [x] Version is updated
- [x] Changed files hash is updated

Azgaar commented:
Hello and thanks for the contribution! It looks interesting and definitely helpful. I can see how the code can be refactored a bit, but will be able to pick it up next week only. Also I am not a fan of the shading change, I believe it looked better before.

yldrefruz commented:
> Hello and thanks for the contribution! It looks interesting and definitely helpful. I can see how the code can be refactored a bit, but will be able to pick it up next week only. Also I am not a fan of the shading change, I believe it looked better before.

I can make the shading change as an option. That way it can make everyone happy. Also can we consider upgrading threejs version. The latest version is 154 and generator currently uses 109. Maybe that way we can add a subdivision utility for 3D Scene (https://github.com/stevinz/three-subdivide/blob/master/example/LoopSubdivision.html current generator's threejs API doesn't have required functions for using something like this.).

Azgaar commented:
Upgrading the version is a good idea. As for the option, it would be better to not overwhelm users with options, especially pure technical ones.

Azgaar commented:
@yldrefruz, what is your plan for this PR?

yldrefruz commented:
I will work on this.

yldrefruz commented:
@Azgaar current shadow map resolution is increased. So this isn't the exactly old rendering method. Should I revert shadow map size?

Note: The smoothness on shadows peaks is because of Subdivision. And subdivision isn't on by default so this won't affect user too much.

Old:
![image](https://github.com/Azgaar/Fantasy-Map-Generator/assets/30903352/bee02990-fa18-47ce-a092-1df643209ecd)


New:
![image](https://github.com/Azgaar/Fantasy-Map-Generator/assets/30903352/50b9fda3-eaf0-41eb-b817-eed2fe68f51f)

Azgaar commented:
Looks fine

Azgaar commented:
Are there any other updates expected?

yldrefruz commented:
No there isn't any other updates related to 3D scene. But maybe i could make some changes on Globe view in another PR if you are okay with.

Azgaar commented:
> No there isn't any other updates related to 3D scene. But maybe i could make some changes on Globe view in another PR if you are okay with.

Yes, why not

Azgaar commented:
Can you do the cleanup (there is some unused commented code and comments) and also check whether everything works fine? I see an error `geometry.getAttribute is not a function`.

yldrefruz commented:
> Can you do the cleanup (there is some unused commented code and comments) and also check whether everything works fine? I see an error `geometry.getAttribute is not a function`.

Where does the error occur? I can't see any error.

yldrefruz commented:
> > Can you do the cleanup (there is some unused commented code and comments) and also check whether everything works fine? I see an error `geometry.getAttribute is not a function`.
> 
> Where does the error occur? I can't see any error. (Sorry i see now)

I see an error but it isn't geometry.getAttribute.

yldrefruz commented:
@Azgaar if there is any error that should be fixed, can you send me the error message, since i can't reproduce that error. The error i get is in graphUtils.js which is an error that happens when you move your mouse while generator is still loading. And i tested the error in 3 different machines. Maybe that error is something about cached files.

Azgaar commented:
The issue occurs on 3D view open on deployed preview only - https://deploy-preview-956--afmg.netlify.app.
There is no issue if I run the code locally.

yldrefruz commented:
> The issue occurs on 3D view open on deployed preview only - https://deploy-preview-956--afmg.netlify.app. There is no issue if I run the code locally.

I opened deploy preview, changed the scene mode to 3D Scene and i couldn't see any error.

Azgaar commented:
Hm, I see it each time. Let's ignore it for now. Generally the addition is very good and I like new functionality added, but I have some minor observations:

* Three.js is a big library and it's used for a very specific case, most users won't need it. Initially it was loaded dynamically, when needed, but now I can see it's added to head to be always loaded. I am not sure why, it increases the network load. Same for the `LoopSubdivision.js`, we can load it together with `orbitControls.min.js`. Also it would be nice to minify it as we do with external libs to save some resources

* I would rename 'Subdivide Model.' to something like 'Smooth geometry' or `Smooth angles` as 'Subdivide Model' is completely unclear for users

* `options3dMeshSkinResolution` option is a bit confusing. You use 1,2,3, etc. values, but then you use switch over it to get size. Why not just use size directly in the select value? I believe it will be better to do the same as already done for the Globe view. Also I see it sets height the same as width, does it distort the texture to make it rectangular?

* For some reason when zoomed out, the next method provides less quality texture even when texture is set to max value:
- before:
<img width="1510" alt="image" src="https://github.com/Azgaar/Fantasy-Map-Generator/assets/26469650/504b7ad8-b448-4979-bd41-686676944d5b">
after:
<img width="1512" alt="image" src="https://github.com/Azgaar/Fantasy-Map-Generator/assets/26469650/112c1bc3-a5ee-4f8f-90d9-6c5c826dcdbd">

* There is some unclear `for3D` option. It looks it is not used

* Wireframe is now commented. If you don't find it useful (as generally I agree here), need to remove the code completely

* There are some unclear comments like `Maybe serialize the models to the .map file`. I don't believe we need them, in any case it's not completely clear what the message is

yldrefruz commented:
- I renamed `for3D` to `for3DRender`. I plan to use this parameter for switching some layers on or off in 3D scene for my next PR. When map uses more screen size than 2D view screen size, the smaller burgs should be visible. So we need another texture with all burgs, etc. visible. This parameter will cause the function to return two different map urls in future. So this is something like a reserved property.
- The skin's width and height is same because `three js` will try to resize it to  power of two. And it will take extra time to resize. And it won't cause any displayed resolution change.
- When you are viewing the plane from distance. The default anisotropy is not enough. I maxed it out.
- Wireframe mode could be useful when exporting the 3D map as obj. I originally planned the export to create a zip file which includes textures for map and original mesh. I planned to make the editor view Wireframe while preparing to export. Additionally wireframe is very useful for debugging since I can call that from console with `ThreeD.toggleWireframe()`
- I moved loading of loopSubdivison to when `SmoothGeometry` option enabled.
- The comment `Maybe serialize the models to the .map file` is the topic of maybe another PR. The message above and in this comment was the custom burg meshes. Which will import OBJ and show the mesh of the burg when `Show 3D Labels` is enabled.

Azgaar commented:
Ok, that's fine. Please fix the version conflict so that I can merge the PR. Please also remove the code that is not used right now, it can be re-added when required.

yldrefruz commented:
Ok I did remove for3DRender.
